### PR TITLE
Improve the logs when the heartbeat response indicates failure

### PR DESCRIPTION
**Pull request by linlinisme**
### Does this pull request fix one issue?

fix https://github.com/alibaba/Sentinel/issues/1194

### CODE REVIEW

**Comment by sczyh30** [comment]
#926 brought some conflicts. Could you please resolve it?

File: sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java [hunk 1]
```diff
@@ -104,7 +106,8 @@ public HttpHeartbeatSender() {
 
     @Override
     public boolean sendHeartbeat() throws Exception {
-        if (StringUtil.isEmpty(consoleHost)) {
+        if (StringUtil.isEmpty(consoleHost) || invalidPort(consolePort)) {
+            RecordLog.warn("[HttpHeartbeatSender] Failed to send heartbeat for invalid args consoleHost:{0} consolePort:{1}", consoleHost, consolePort);
```
Comment (sczyh30) [inline sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: Maybe it's not necessary to log here, as we also checked and warned in the constructor of the HeartbeatSender impls.

File: sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java [hunk 2]
```diff
@@ -89,11 +91,43 @@ public boolean sendHeartbeat() throws Exception {
         // Send heartbeat request.
         CloseableHttpResponse response = client.execute(request);
         response.close();
-        return true;
+        int statusCode = response.getStatusLine().getStatusCode();
+        if (statusCode == OK_STATUS) {
+            return true;
+        } else if (clientErrorCode(statusCode) || serverErrorCode(statusCode)) {
+            RecordLog.warn("[HttpHeartbeatSender] Failed to send heartbeat to "
+                    + consoleHost + ":" + consolePort + ", response : ", response);
```
[the remark this answers is not in the document]
Comment (sczyh30) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: Yes. Maybe just logging the response status code is enough?

[the remark this answers is not in the document]
Comment (linlinisme) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: emm. I check  the response Headers. It is a supplement of the server version、location 、name, may be it is not necessary to print.

**Comment by linlinisme** [comment]
friendly ping :)

File: sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (cdfive) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: How about else branch, do we need to add some log?  The **NPE** remains in `SimpleHttpResponse#getBodyAsString()` at line 108, when body is null it throws.

[the remark this answers is not in the document]
Comment (linlinisme) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: Let me fix it together, it is not complicated

[the remark this answers is not in the document]
Comment (linlinisme) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: any suggestion?

[the remark this answers is not in the document]
Comment (cdfive) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: I think that 3 need to check.To verify it, we can throw Exception in `MachineRegistryController#receiveHeartBeat` of dashboard, and call `SimpleHttpResponse#toString()`.

[the remark this answers is not in the document]
Comment (linlinisme) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: > I think that 3 need to check.To verify it, we can throw Exception in `MachineRegistryController#receiveHeartBeat` of dashboard, and call `SimpleHttpResponse#toString()`.

It still will return http 200 ok response code, even though  throw ex.  It seems `SimpleHttpResponseParser`  don't check the response data, which may  need to be checked. How about we open another issue to discuss ?

[the remark this answers is not in the document]
Comment (sczyh30) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: > > I think that 3 need to check.To verify it, we can throw Exception in `MachineRegistryController#receiveHeartBeat` of dashboard, and call `SimpleHttpResponse#toString()`.
> 
> It still will return http 200 ok response code, even though throw ex. It seems `SimpleHttpResponseParser` don't check the response data, which may need to be checked. How about we open another issue to discuss ?

Yeah, we could open another issue to discuss how to improve it :)

[the remark this answers is not in the document]
Comment (cdfive) [reply sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: Yes, one issue one PR is clean and consistent.

> It still will return http 200 ok response code, even though throw ex. It seems SimpleHttpResponseParser don't check the response data
I got http 500 and the `SimpleHttpResponseParser` checked the response.

Add this in `MachineRegistryController#receiveHeartBeat`
```java
if (true) {
    throw new RuntimeException("test");
}
```
Then add in `SimpleHttpHeartbeatSender#sendHeartbeat`
```java
try {
    System.out.println(response);
} catch (Exception e) {
    e.printStackTrace();
}
```
Then the response code 500 got, and **NPE** occured.

File: sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java [hunk 3]
```diff
@@ -89,11 +91,43 @@ public boolean sendHeartbeat() throws Exception {
...
+        if (statusCode == OK_STATUS) {
+            return true;
+        } else if (clientErrorCode(statusCode) || serverErrorCode(statusCode)) {
+            RecordLog.warn("[HttpHeartbeatSender] Failed to send heartbeat to "
+                    + consoleHost + ":" + consolePort + ", http status code: {0}", statusCode);
+        }
+
+        return false;
+
+
```
Comment (cdfive) [inline sentinel-transport/sentinel-transport-netty-http/src/main/java/com/alibaba/csp/sentinel/transport/heartbeat/HttpHeartbeatSender.java]: Do we need close the response manually,  `response.close();`

**Comment by sczyh30** [comment]
Thanks!

**Pull request merged**